File: DailyChallenges/Scala/LeetCode_1406.scala
object Solution {
    def stoneGameIII(stoneValue: Array[Int]): String = {
        val dp = Array.fill[Int](4)(0)
        for (i<- (stoneValue.length-1) to 0 by -1){
            dp(i%4) = Int.MinValue;
            var taken = 0;
            for (j <- 0 to 2 if i+j < stoneValue.length){
                taken += stoneValue(i+j)
                dp(i%4) = dp(i%4) max (taken - dp((i+j+1)%4))
            }
        }
        dp(0) match {
            case 0 => "Tie"
            case v if v>0 => "Alice"
            case _ => "Bob"
        }
    }
}

File: DailyChallenges/CPP/LeetCode_1406.cpp
class Solution {
public:
    string stoneGameIII(vector<int>& stoneValue) {
        vector<int> dp(4); // backwards only 3
        int n = stoneValue.size();
        for (int i = n-1; i >= 0; --i){ // O(n) from back to front
            dp[i%4] = INT_MIN;
            for (int j = 0, taken = 0; j<3 && i+j<n; ++j) // O(3)
                dp[i%4] = max(dp[i%4], (taken += stoneValue[i+j]) - dp[(i+j+1)%4]); 
        }
        return dp[0] == 0 ? "Tie" : (dp[0] > 0 ? "Alice" : "Bob");
    }
};
